assignment/final/aLife/code/master.js

```diff
@@ -0,0 +1,417 @@
+let selectedType1;
+let selectedType2;
+
+pkmnListDiv.innerHTML = "Loading Pokemon";
+
+function fetchPkmn() {
+    fetch(baseUrl + 'pokedex/1/')
+        .then(res => res.json())
+        .then(data => {
+            for (let i = 0; i < data.pokemon_entries.length; i++) {
+                pkmnList.push({
+                    ID: data.pokemon_entries[i].entry_number,
+                    NAME: data.pokemon_entries[i].pokemon_species.name,
+                    PKMNURL: data.pokemon_entries[i].pokemon_species.url
+                });
+            }
+            pkmnListDiv.innerHTML = "";
+            for (let i = 0; i < pkmnList.length; i++) {
+                pkmnListDiv.innerHTML += `<p id="${pkmnList[i].NAME}">${pkmnList[i].NAME}</p>`;
+            }
+            //console.log(pkmnList);
+        });
+}
+
+function fetchTypes() {
+    fetch(baseUrl + 'type/')
+        .then(res => res.json())
+        .then(data => {
+            for (let i = 0; i < data.results.length; i++) {
+                pkmnTypes.push({
+                    TYPE: data.results[i].name,
+                    TYPEURL: data.results[i].url
+                });
+            }
+            //console.log(pkmnTypes);
+            for (let i = 0; i < pkmnTypes.length; i++) {
+                typesCorespondingPkmn.push({
+                    TYPE: pkmnTypes[i],
+                    LIST: []
+                });
+
+                let urlForList = pkmnTypes[i].TYPEURL;
+                fetch(urlForList)
+                    .then(res => res.json())
+                    .then(data => {
+                        //console.log(data.pokemon);
+                        typesCorespondingPkmn[i].LIST = data.pokemon;
+                    });
+            }
+            //console.log(typesCorespondingPkmn);
+        });
+    //console.log(moves);
+    //console.log(pkmnTypes);
+}
+
+function fetchMoves() {
+    fetch(baseUrl + 'move/')
+        .then(res => res.json())
+        .then(data => {
+            let num = data.count;
+            fetch(baseUrl + 'move?offset=0&limit=' + num)
+                .then(res => res.json())
+                .then(data => {
+                    for (let i = 0; i < num; i++) {
+                        fetch(data.results[i].url)
+                            .then(res => res.json())
+                            .then(data => {
+                                moves.push({
+                                    NAME: data.name,
+                                    TYPE: data.type.name,
+                                    POWER: data.power
+                                });
+                            });
+                    }
+                });
+        });
+}
+
+function getCurrPkmnMoves(pkmn, data) {
+    moveset.innerHTML = "";
+    currentMoves = [];
+    for (let i = 0; i < data.moves.length; i++) {
+        let move = moves.find(name => name.NAME === data.moves[i].move.name);
+        moveset.innerHTML +=
```

Comment:
Just a heads up on this part.  Typically we want to touch the DOM as little as possible...that means when we make changes we want to kind of batch them and give them in a meaningful way.  This is important for performance and for the actually layout/drawing of the page.  

In this case `moveset` refers to an actual element that you are constantly going and updating with data and causing the page to try and reflow/draw/color the page.  A lot of times you will want to get your changes together and the set in once and as few times as possible.  

Refactoring it would be along the lines of building the whole string for html first and then only calling it once:
```
move.innerHTML = myEntireChunkOfHTML
```

Reply:
Oh yeah.. I think you showed us that, I just didn’t think about it I guess.